File: joiner-core/src/main/java/cz/encircled/joiner/util/JoinerUtil.java
package cz.encircled.joiner.util;

import java.util.LinkedList;
import java.util.List;
import java.util.Set;

import com.mysema.query.types.EntityPath;
import cz.encircled.joiner.query.join.JoinDescription;

/**
 * @author Kisel on 25.01.2016.
 */
public class JoinerUtil {

    public static List<JoinDescription> unrollChildrenJoins(Set<JoinDescription> joins) {
        List<JoinDescription> collection = new LinkedList<>();

        for (JoinDescription joinDescription : joins) {
            unrollChildrenInternal(joinDescription, collection);
        }

        return collection;
    }

    private static void unrollChildrenInternal(JoinDescription join, List<JoinDescription> collection) {
        collection.add(join);
        for (JoinDescription child : join.getChildren()) {
            unrollChildrenInternal(child, collection);
        }
    }

    public static <T extends EntityPath> T getAliasForChild(EntityPath<?> parent, T childPath) {
        return ReflectionUtils.instantiate(childPath.getClass(), childPath.toString() + "_on_" + parent.toString());
    }

}

File: joiner-core/src/main/java/cz/encircled/joiner/util/JoinerUtil.kt
package cz.encircled.joiner.util

import com.mysema.query.types.EntityPath
import cz.encircled.joiner.exception.JoinerException

/**
 * @author Kisel on 25.01.2016.
 */
object JoinerUtil {

    @SuppressWarnings("unchecked")
    fun <T : EntityPath<*>> instantiate(generatedClass: Class<out EntityPath<*>>, alias: String): T {
        Assert.notNull(alias)

        try {
            val constructor = generatedClass.getConstructor(String::class.java)
            return constructor.newInstance(alias) as T
        } catch (e: NoSuchMethodException) {
            throw JoinerException("EntityPath String constructor is missing on " + generatedClass)
        } catch (e: Exception) {
            throw JoinerException("Failed to create new instance of " + generatedClass, e)
        }

    }

    fun <T : EntityPath<*>> getAliasForChild(parent: EntityPath<*>, childPath: T): T {
        return JoinerUtil.instantiate<T>(childPath.javaClass, childPath.toString() + "_on_" + parent.toString())
    }

}
